Provide a new/branch which updates PR(856) for including the dh5walk …

rawarren opened the pull request:
This replaces the old PR#856.  It seemed more convenient to start with 'develop' and then add the new code into the utils directory.

byrnHDF commented on Makefile.am:
```diff
@@ -88,9 +88,9 @@ else
 endif
 
-SUBDIRS = src $(TESTSERIAL_DIR) $(TESTPARALLEL_DIR) bin utils $(TOOLS_DIR) . \
+SUBDIRS = src $(TESTSERIAL_DIR) $(TESTPARALLEL_DIR) bin $(TOOLS_DIR) utils . \
```
We need to make sure but I think utils is an optional component.

rawarren replied:
I'm only changing the order of when the utils directory gets traversed..  Am I missing something in whether the order is important, i.e. I wanted to make sure that utils got visited AFTER the tools.   I should probably add a conditional statement to prevent dh5walk from be built if we don't build tools...

lrknox replied:
> 
> 
> I'm only changing the order of when the utils directory gets traversed.. Am I missing something in whether the order is important, i.e. I wanted to make sure that utils got visited AFTER the tools. I should probably add a conditional statement to prevent dh5walk from be built if we don't build tools...

Assuming mirror_vfd should always be built, with or without tools, as long as the conditional only applies to dh5walk, I suppose if tools aren't built, then utils/mirror_vfd will be built next after bin.

byrnHDF replied:
I'm thinking utils is an optional component therefore you would need to use a variable -> $(UTILS_DIR)
And anyway, if tools is not built, can anything in utils be built? Still I think utils is optional.

rawarren replied:
I added a conditional check the utils/Makefile.am
if BUILD_TOOLS_CONDITIONAL
   TOOLS_DIR =tools
else
   TOOLS_DIR=
endif

byrnHDF commented on release_docs/RELEASE.txt:
```diff
@@ -47,6 +47,12 @@ New Features
     -------------
+    - Added new configure option to support building parallel tools.
+      See Tools below (autotools - CMake):
+            --with-libmfu=DIR      HDF5_BUILD_PARALLEL_TOOLS
```
the autotools name is very different ten the CMake one?

rawarren replied:
Yeah... The CMake version indirectly uses FindMFU.cmake, etc. to hopefully locate the necessary libraries and include directories so that dh5walk can be built.   I'm happy to make these more similar, do you have a suggestion?

byrnHDF commented on utils/Makefile.am:
```diff
@@ -33,6 +33,6 @@ else
 
 # All subdirectories
-SUBDIRS=$(MIRROR_VFD_DIR) $(TESTUTIL_DIR)
+SUBDIRS=$(MIRROR_VFD_DIR) $(TESTUTIL_DIR) tools
```
is tools optional in autoconf - the it should have a variable assigned ($(TOOLSUTIL_DIR)?)

rawarren replied:
This was fixed by adding the conditional if BUILD_TOOLS_CONDITIONAL, so tools is now $(TOOLS_DIR)

byrnHDF commented on utils/tools/Makefile.am:
```diff
@@ -0,0 +1,38 @@
+CONFIG=ordered
+
+# All subdirectories
+SUBDIRS=$(DH5WALK) $(TESTSERIAL_DIR)
```
CMake has dh5walk subdirectory as absolute - should that be changed to match the optional condition here?

rawarren replied:
Fixed

rawarren committed:
Provide a new/branch which updates PR(856) for including the dh5walk parallel tools.

byrnHDF commented on utils/tools/dh5walk/Makefile:
```diff
@@ -0,0 +1,1660 @@
+# Makefile.in generated by automake 1.16.1 from Makefile.am.
```
I don't think this file gets committed - it is generated from the ".am" file.

rawarren replied:
Deleted this file

byrnHDF commented on utils/tools/dh5walk/Makefile.am:
```diff
@@ -0,0 +1,47 @@
+# Include src directory
+AM_CPPFLAGS+=-I$(top_srcdir)/src -I$(top_srcdir)/tools/lib
+
+TOOLSOURCES=h5tools.c h5tools_dump.c h5tools_str.c h5tools_utils.c h5diff.c \
```
This is rebuilding the tools library - should just be the dh5walk tool, right?

rawarren replied:
Fixed, i.e. removed TOOLSOURCES and I'm just linking against $LIBH5TOOLS

byrnHDF commented on utils/tools/test/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+project (HDF5_TOOLS_TEST C)
+
+#-- Add the h5diff tests
+add_subdirectory (dh5walk)
```
Is it possible to be here without having enabled dh5walk tool?  See the Makefile.am file where this is optional.

rawarren replied:
Fixed:
project (HDF5_TOOLS_TEST C)

#-- Add the h5diff tests                                                                                                                                                                                                                
if (HDF5_BUILD_PARALLEL_TOOLS)
   add_subdirectory (dh5walk)
endif()

byrnHDF commented on utils/tools/test/dh5walk/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+cmake_minimum_required (VERSION 3.12)
+project (HDF5_TOOLS_TEST_DH5WALK)
+
+add_custom_command(
```
Should this be an optional command (add an if block?)

rawarren replied:
Added the if (HDF5_BUILD_PARALLEL_TOOLS)  block...

byrnHDF commented on utils/tools/test/dh5walk/CMakeTests.cmake:
```diff
@@ -0,0 +1,162 @@
+  # --------------------------------------------------------------------
+  # Copy all the test files from source directory to test directory
+  # --------------------------------------------------------------------
+  set (LIST_HDF5_TEST_FILES
```
The plan is to test all these files?

rawarren replied:
Eventually.   The problem in the short term is to get the matching .txt files to validate the results.   Ideally we reuse the testing validations that already exist for h5diff and h5ls, but that will take a bit more time

byrnHDF replied:
But do we want to test all the files? We should be testing dh5walk not the functionality of h5diff/h5ls. I would choose just a couple of test files, ones that might expose a dh5walk problem, to use for testing. The tools would already have been tested.

rawarren committed:
Fix various bugs/issues identified during review and further testing

rawarren committed:
Updating the repo with additional fixes the result from the PR comments

rawarren committed:
Merge branch 'develop' of https://github.com/HDFGroup/hdf5 into feature/PR856

rawarren committed:
Deleted Makefile which was inadvertently included in one of the previous commits.

rawarren committed:
Merge branch 'develop' of https://github.com/HDFGroup/hdf5 into feature/PR856

rawarren committed:
Fixed the last of the identified issues with the previous PR issues

rawarren committed:
Merge branch 'develop' of https://github.com/HDFGroup/hdf5 into feature/PR856

byrnHDF reviewed (approved):
Let's give it a try - probably not till Monday because of the other merges Larry is planning.

rawarren commented:
Thanks Allen!
Monday should be fine...
-Richard

> On Nov 12, 2021, at 4:17 PM, Allen Byrne ***@***.***> wrote:
> 
> 
> @byrnHDF approved this pull request.
> 
> Let's give it a try - probably not till Monday because of the other merges Larry is planning.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/HDFGroup/hdf5/pull/1177#pullrequestreview-805233559>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AC5PMX5QEZHU2MUDT4NI3KDULV76BANCNFSM5HWJWDEQ>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
>

lrknox commented on MANIFEST:
```diff
@@ -2995,6 +2999,23 @@
+./utils/tools/CMakeLists.txt
+./utils/tools/Makefile.am
+./utils/tools/dh5walk/CMakeLists.txt
+./utils/tools/dh5walk/Makefile
```
Makefile is not a source file:  should be removed from MANIFEST

rawarren replied:
Fixed in my latest commit..  Thank you!

rawarren committed:
Merge branch 'develop' of https://github.com/HDFGroup/hdf5 into feature/PR856

rawarren committed:
Removed utils/tools/dh5walk/Makefile from the MANIFEST file since this is a derived file

github-actions[bot] committed:
Committing clang-format changes

lrknox commented:
One of the checks failed, I think because in autotools configure:  "configure: error: could't find libfmu library".  The other 9 tests were then cancelled.  See https://github.com/HDFGroup/hdf5/actions/runs/1456574636.  The check results don't show up on this page because clang-format reformatted the code, and then the results can be found only under Actions.  The workaround to make them show up here is to load module LLVM on jelly to get clang version 10.1.0, run bin/format_source in the top-level source directory, and commit the resulting code.

rawarren committed:
Updated configure.ac to reintroduce --with-libmfu=DIR as an option for locating the necessary libraries and include files for dh5walk

rawarren committed:
Updates to dh5walk logfile generation.  Will test this on jelly to verify that it works correctly

github-actions[bot] committed:
Committing clang-format changes

rawarren committed:
Made changes to configure.ac based on Larry's comments on how to remove -lmfu from the actual libhdf5 library.

rawarren committed:
Merge branch 'feature/PR856' of https://github.com/HDFGroup/hdf5 into feature/PR856

Commit:
Various changes in configure.ac to allow AM_LDFLAGS to be set, but exclude linking in -lmfu for libhdf5

github-actions[bot] committed:
Committing clang-format changes

rawarren committed:
Fixed the testing to simply check that dh5walk logs are non-zero size and/or match an expected size.

lrknox committed:
Modified configure and dh5walk/Makefil.am and copy scripts for autotools
configure.

rawarren committed:
Merge pull request #1218 from lrknox/atconfigure-PR856

Modified configure and dh5walk/Makefil.am and copy scripts for autotools

rawarren committed:
Finish renaming of dh5walk->h5dwalk along with documentation and build changes for autotools and CCMake.

rawarren committed:
Various updates pertaining to the rename of dh5walk -> h5dwalk and testing (both autotools and CMake)

lrknox committed:
Make --with-libmfu= option dependent on --enable-parallel-tools and
--enable-parallel in configure.ac.
Change remaining dh5walk=>h5dwalk ad DH5WALK=>H5DWALK.

byrnHDF commented on utils/tools/test/h5dwalk/copy_demo_files.sh.in:
```diff
@@ -0,0 +1,86 @@
+# If you do not have access to either file, you may request a copy from
+# help@hdfgroup.org.
+#
+srcdir=@srcdir@
```
This variable should be set in a configure_file command before use.

byrnHDF commented on utils/tools/test/h5dwalk/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+project (HDF5_TOOLS_TEST_DH5WALK)
+
+if (HDF5_BUILD_PARALLEL_TOOLS)
+  add_custom_command(
```
Do we really want to always run this command - shouldn't it be part of the test/run script? At least it should be a prerequisite add_test and not a custom command.
Also, this is where the configure file should be called (Like in test/ShellTests.cmake);
  find_program (SH_PROGRAM bash)
  if (SH_PROGRAM)
    set (srcdir ${HDF5_TEST_SOURCE_DIR})
    configure_file(${HDF5_TEST_SOURCE_DIR}/test_usecases.sh.in 
      ${HDF5_TEST_BINARY_DIR}/H5TEST/test_usecases.sh @ONLY)
...
endif ()

byrnHDF commented on utils/tools/test/h5dwalk/CMakeTests.cmake:
```diff
@@ -0,0 +1,58 @@
+              -D "TEST_OUTPUT=${resultfile}.out"
+              -D "TEST_EXPECT=${resultcode}"
+              -D "TEST_REFERENCE=${resultfile}.h5dwalk"
+              -P "${HDF_RESOURCES_EXT_DIR}/runTest.cmake"
```
Add the -D "TEST_LIBRARY_DIRECTORY=${LL_PATH}" in this list.

byrnHDF replied:
The question still is where else is this ENV variable being used.  ${LL_PATH} should be available here if it is available in the root CMakeLists.txt file.

byrnHDF commented on utils/tools/test/h5dwalk/CMakeTests.cmake:
```diff
@@ -0,0 +1,58 @@
+
+#  get_property(_h5dwalk_llenv GLOBAL PROPERTY H5DWALK_ENVIRONMENT)
+  message(STATUS "H5DWALK_TEST_ENV = $ENV{H5DWALK_TEST_ENV}")
+  set(LD_LIBRARY_PATH "ENV{LD_LIBRARY_PATH}:$ENV{H5DWALK_TEST_ENV}")
```
Remove this.

byrnHDF commented on CMakeLists.txt:
```diff
@@ -1008,6 +1008,32 @@ endif ()
+    find_package(MFU REQUIRED)
+    if (MFU_FOUND)
+      message(STATUS "LL_PATH=${LL_PATH}")
+      set (ENV{H5DWALK_TEST_ENV} "${LL_PATH}")
```
Where is H5DWALK_TEST_ENV used?

lrknox replied:
It's used now in utils/tools/test/h5dwalk/CMakeTests.cmake.  It provides the library path for TEST_LIBRARY_DIRECTORY.  (It works for me, and presumably for Richard, too.)

byrnHDF replied:
However if the suggested changes are made - this should be unnecessary.

lrknox replied:
Do you mean adding -D "TEST_LIBRARY_DIRECTORY=${LL_PATH}" to CMakeTests.cmake?  When I did that, the result in build/utils/tools/test/h5dwalk/CTestTestfile.cmake was "-D" "TEST_LIBRARY_DIRECTORY=", but with -D "TEST_LIBRARY_DIRECTORY=$ENV{H5DWALK_TEST_ENV}" it is "-D" "TEST_LIBRARY_DIRECTORY=/mnt/hdf/packages/mpifileutils/install/lib64:/mnt/hdf/packages/mpifileutils/install/lib".

rawarren committed:
Updates to allow LD_LIBRARY_PATH to be set for testing

byrnHDF commented on CMakeLists.txt:
```diff
@@ -1008,6 +1008,32 @@ endif ()
+    set (CMAKE_PREFIX_PATH "$HDF_RESOURCES_DIR")
+    find_package(MFU REQUIRED)
+    if (MFU_FOUND)
+      message(STATUS "LL_PATH=${LL_PATH}")
```
If this works here, then LL_PATH should be available in the test.

byrnHDF commented on CMakeLists.txt:
```diff
@@ -1008,6 +1008,32 @@ endif ()
 #-----------------------------------------------------------------------------
 if (EXISTS "${HDF5_SOURCE_DIR}/utils" AND IS_DIRECTORY "${HDF5_SOURCE_DIR}/utils")
   add_subdirectory (utils)
```
Move this add_directory to the bottom of the if block.

rawarren committed:
Remove the unnecessary quotes when using @LL_PATH@

lrknox commented:
Yes, with that moved to the bottom, LL_PATH is available for utils/tools/test/h5dwalk/CMakeTests.cmake, and H5DWALK_TEST_ENV is not needed.


From: Allen Byrne ***@***.***>
Sent: Monday, November 29, 2021 10:44 AM
To: HDFGroup/hdf5 ***@***.***>
Cc: Larry Knox ***@***.***>; Review requested ***@***.***>
Subject: Re: [HDFGroup/hdf5] Provide a new/branch which updates PR(856) for including the dh5walk … (PR #1177)


@byrnHDF commented on this pull request.

________________________________

In CMakeLists.txt<https://github.com/HDFGroup/hdf5/pull/1177#discussion_r758545731>:

> @@ -1008,6 +1008,32 @@ endif ()

 #-----------------------------------------------------------------------------

 if (EXISTS "${HDF5_SOURCE_DIR}/utils" AND IS_DIRECTORY "${HDF5_SOURCE_DIR}/utils")

   add_subdirectory (utils)

Move this add_directory to the bottom of the if block.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://github.com/HDFGroup/hdf5/pull/1177#pullrequestreview-818051133>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ADGMWCUBVD7FY2PBXLT4GUTUOOUW3ANCNFSM5HWJWDEQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

lrknox committed:
Merge commit '263ebcc7b5759631c1a717a6860905cd2576f025' into HEAD

rawarren committed:
Fixed some missing file changes resulting from the dh5walk -> h5dwalk update

lrknox committed:
Merge branch 'feature/PR856' into atconfigure-PR856

rawarren committed:
Merge pull request #1233 from lrknox/atconfigure-PR856

Atconfigure pr856

lrknox committed:
Remove unnecessary CMake variable H5DWALK_TEST_ENV.
